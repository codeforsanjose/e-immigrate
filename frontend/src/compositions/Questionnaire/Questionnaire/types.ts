import { ReactSetter } from "../../../types/common";

export type OtherComponentProps = {
    setShowFollowUp: ReactSetter<Record<string, boolean>>;
    setErrors: ReactSetter<Record<string, unknown>>;
};

export type FollowupMap = Record<string, boolean>;
