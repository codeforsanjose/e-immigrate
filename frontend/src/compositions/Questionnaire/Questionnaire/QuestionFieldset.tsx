import React from 'react';

import { OtherComponentProps } from './types';
import { FormElementWrapper } from '../../../components/FormElementWrapper';
import { useContentContext } from '../../../contexts/ContentContext';
import { QuestionInfo } from '../../../types/ApiResults';

type QuestionFieldsetProps = {
    question: QuestionInfo;
    others: OtherComponentProps;
};
export function QuestionFieldset(props: QuestionFieldsetProps) {
    const {
        question,
        others,
    } = props;
    const { content } = useContentContext();
    return (
        <fieldset className="Question">
            <div className="QuestionText">
                {question.text}
                {(question.required ?? false)
                    ? ` (${content.required})`
                    : ` (${content.optional})`}
            </div>
            <FormElementWrapper
                elementName={question.questionType}
                others={others}
                question={question}
            />
        </fieldset>
    );
}