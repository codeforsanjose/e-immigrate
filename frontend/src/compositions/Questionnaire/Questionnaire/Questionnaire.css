.Questionnaire {
    max-width: 600px;
    margin: auto;
    display: grid;
}

.Questionnaire button {
    margin: 48px auto 16px;
}

.Questionnaire .Question {
    border: none;
    padding: 1rem;
}

.Questionnaire .FollowUp {
    margin-left: 2rem;
    display: grid;
    grid-template-columns: 30px auto;
}

.Questionnaire .Question .QuestionText {
    font-size: 1rem;
    font-weight: 600;
    margin-bottom: 0.5rem;
}

.Questionnaire .QuestionInput:invalid {
    border: 2px solid var(--color-secondary);
}

.Questionnaire .Question .RequiredError {
    display: none;
    color: var(--color-secondary);
}

.Questionnaire .TextInput:invalid[data-touched='true'] ~ .RequiredError {
    display: block;
}
