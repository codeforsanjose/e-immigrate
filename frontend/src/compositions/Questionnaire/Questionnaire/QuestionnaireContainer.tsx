import React from 'react';
type QuestionnaireContainerProps = {
    children: React.ReactNode;
};
export function QuestionnaireContainer(props: QuestionnaireContainerProps) {
    const {
        children,
    } = props;
    return (
        <div className='Questionnaire'>
            {children}
        </div>
    );
}