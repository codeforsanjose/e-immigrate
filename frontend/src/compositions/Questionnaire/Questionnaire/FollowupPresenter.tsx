/* eslint-disable @typescript-eslint/no-unused-vars */
import React from 'react';

import { QuestionFieldset } from './QuestionFieldset';
import { FollowupMap, OtherComponentProps } from './types';
import Arrow from '../../../data/images/Arrow-Down-Right.svg';
import { QuestionInfo } from '../../../types/ApiResults';
type FollowupPresenterProps = {
    showFollowup: FollowupMap;
    question: QuestionInfo;
    filteredQuestions: Array<QuestionInfo>;
    followUpQuestions: Array<QuestionInfo>;
    others: OtherComponentProps;
};
export function FollowupPresenter(props: FollowupPresenterProps) {
    const {
        showFollowup,
        question,
        filteredQuestions,
        followUpQuestions,
        others,
    } = props;
    const show = showFollowup[question.slug];
    if (!show) return null;
    return (
        <>
            {followUpQuestions.map(followUpQuestion => {
                return (
                    <div
                        className="FollowUp"
                        key={followUpQuestion.slug}
                    >
                        <Arrow
                            height="36px"
                            width="36px" 
                        />
                        <QuestionFieldset
                            others={others}
                            question={followUpQuestion}
                        />
                    </div>
                );
            })}
        </>
    );
}