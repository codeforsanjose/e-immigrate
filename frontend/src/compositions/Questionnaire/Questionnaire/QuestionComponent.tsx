import React from 'react';
import { Question } from '../../../types/Question';
import { FollowupPresenter } from './FollowupPresenter';
import { QuestionFieldset } from './QuestionFieldset';
import { FollowupMap, OtherComponentProps } from './types';

type QuestionComponentProps = {
    showFollowUp: FollowupMap;
    filteredQuestions: Array<Question>;
    parent?: Question;
    question: Question;
    others: OtherComponentProps;
};
export function QuestionComponent(props: QuestionComponentProps) {
    const {
        filteredQuestions,
        showFollowUp,
        question,
        others,
        parent,
    } = props;

    const parentSlug = question.parentQuestionSlug;
    if (!(parentSlug == null || parentSlug === '' || (showFollowUp[parentSlug] && parent?.slug === parentSlug))) return null;
    const followUpQuestions = filteredQuestions.filter(q => q.parentQuestionSlug === question.slug);
    return (
        <div key={question.slug}>
            <QuestionFieldset
                others={others}
                question={question}
            />
            
            <FollowupPresenter
                showFollowup={showFollowUp}
                filteredQuestions={filteredQuestions}
                question={question}
                followUpQuestions={followUpQuestions}
                others={others}
            />
        </div>
    );
}