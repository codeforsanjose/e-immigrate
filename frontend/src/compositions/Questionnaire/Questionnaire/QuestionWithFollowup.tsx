import React from 'react';

import { FollowupPresenter } from './FollowupPresenter';
import { QuestionFieldset } from './QuestionFieldset';
import { FollowupMap, OtherComponentProps } from './types';
import { QuestionInfo } from '../../../types/ApiResults';

type QuestionWithFollowupProps = {
    showFollowUp: FollowupMap;
    filteredQuestions: Array<QuestionInfo>;
    parent?: QuestionInfo;
    question: QuestionInfo;
    others: OtherComponentProps;
};

function canRender(config: {
    showFollowUp: FollowupMap;
    question: QuestionInfo;
    parent?: QuestionInfo;
}) {
    const {
        question,
        showFollowUp,
        parent,
    } = config;

    const parentSlug = question.parentQuestionSlug;
    // if the question isnt dependent on a parent, then render
    if (parentSlug == null || parentSlug === '') return true;
    // if the dependent question isnt triggered, DONT render
    else if (!showFollowUp[parentSlug]) return false;
    // render if we are rendering as a child of the proper parent
    return parent?.slug === parentSlug;
}



export function QuestionWithFollowup(props: QuestionWithFollowupProps) {
    const {
        filteredQuestions,
        showFollowUp,
        question,
        others,
        parent,
    } = props;

    if (!canRender({
        parent,
        question,
        showFollowUp,
    })) return null;

    const followUpQuestions = filteredQuestions.filter(q => q.parentQuestionSlug === question.slug);
    return (
        <div key={question.slug}>
            <QuestionFieldset
                others={others}
                question={question}
            />
            
            <FollowupPresenter
                showFollowup={showFollowUp}
                filteredQuestions={filteredQuestions}
                question={question}
                followUpQuestions={followUpQuestions}
                others={others}
            />
        </div>
    );
}