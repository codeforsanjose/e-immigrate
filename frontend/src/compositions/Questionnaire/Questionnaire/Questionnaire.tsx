import React from 'react';
import { Button } from '../../../components/Button/Button';
import { useMarkFieldAsTouched } from '../hooks/useMarkFieldAsTouched';
import { QuestionnaireIntro } from '../QuestionnaireIntro/QuestionnaireIntro';
import Arrow from '../../../data/images/Arrow-Down-Right.svg';

import './Questionnaire.css';
import { CollectAnswerFunction } from '../../../types/common';
import { ContentText } from '../../../types/ContentText';
import { Question } from '../../../types/Question';
import { FormElementWrapper } from '../../../components/FormElementWrapper';

export type QuestionnaireResponse = Record<string, unknown>;

type QuestionnaireProps = {
    questions: Array<Question>;
    submitQuestionnaireResponse: (value: QuestionnaireResponse) => void;
    questionnaireResponse: QuestionnaireResponse;
    setQuestionnaireResponse: (value: QuestionnaireResponse) => void;
    content: ContentText;
    collectAnswer: CollectAnswerFunction;
};

export function Questionnaire(props: QuestionnaireProps) {
    const {
        collectAnswer,
        content,
        questionnaireResponse,
        questions,
        submitQuestionnaireResponse,
    } = props;
    const categories = ['Basic Info', 'Waiver Flag', 'Red Flag'];
    const [categoryIndex, setCategoryIndex] = React.useState(0);
    const {
        bindField,
        setAllFieldsTouched,
    } = useMarkFieldAsTouched();
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const [errors, setErrors] = React.useState<Record<string, unknown>>({});
    const [introPage, setIntroPage] = React.useState(true);
    const [showFollowUp, setShowFollowUp] = React.useState<Record<string, boolean>>({});

    const filteredQuestions = questions.filter(
        (q) => q.category === categories[categoryIndex],
    );
    const formElementWrapperOthers = React.useMemo(() => {
        return {
            bindField,
            collectAnswer,
            setShowFollowUp,
            setErrors,
        };
    }, [bindField, collectAnswer]);
    const onSubmit = () => {
        submitQuestionnaireResponse(questionnaireResponse);
    };
    const nextStep = (e: {
        preventDefault: () => void;
    }) => {
        e.preventDefault();
        const allRequiredFieldsCompleted = filteredQuestions.every((q) => {
            if ((q.required ?? false) && !(Boolean(questionnaireResponse[q.slug]))) {
                if (q.parentQuestionSlug != null) {
                    if (questionnaireResponse[q.parentQuestionSlug] === 'Yes') {
                        return false;
                    }
                    else {
                        return true;
                    }
                }
                return false;
            }
            else {
                return true;
            }
        });
        setAllFieldsTouched();
        if (allRequiredFieldsCompleted) {
            if (categoryIndex < categories.length - 1) {
                setCategoryIndex((prev) => prev + 1);
            }
            else {
                onSubmit();
                return;
            }
        }
        else {
            alert(`Please complete every question`);
        }
    };
    return (
        <div className="Questionnaire">
            {introPage
                ? (
                    <QuestionnaireIntro
                        content={content}
                        setIntroPage={setIntroPage} />
                )
                : (
                    <>
                        <>
                            {filteredQuestions.map((question) => {
                                const followUpQuestions = filteredQuestions.filter(
                                    (q) => q.parentQuestionSlug === question.slug,
                                );
                                if (question.parentQuestionSlug != null) {
                                    return null;
                                }
                                return (
                                    <div key={question.slug}>
                                        <fieldset className="Question">
                                            <div className="QuestionText">
                                                {question.text}
                                                {(question.required ?? false)
                                                    ? ` (${content.required})`
                                                    : ` (${content.optional})`}
                                            </div>
                                            <FormElementWrapper
                                                elementName={question.questionType}
                                                content={content}
                                                others={formElementWrapperOthers}
                                                question={question}
                                            />
                                        
                                        </fieldset>
                                        {showFollowUp[question.slug] &&
                                        followUpQuestions.map(
                                            (followUpQuestion) => {
                                                return (
                                                    <div
                                                        className="FollowUp"
                                                        key={followUpQuestion.slug}
                                                    >
                                                        <Arrow
                                                            height="36px"
                                                            width="36px" />
                                                        <fieldset className="Question">
                                                            <div className="QuestionText">
                                                                {followUpQuestion.text}
                                                                {(followUpQuestion.required ?? false)
                                                                    ? ` (${content.required})`
                                                                    : ` (${content.optional})`}
                                                            </div>
                                                            <FormElementWrapper
                                                                elementName={followUpQuestion.questionType}
                                                                content={content}
                                                                others={formElementWrapperOthers}
                                                                question={followUpQuestion}
                                                            />
                                                        </fieldset>
                                                    </div>
                                                );
                                            },
                                        )}
                                    </div>
                                );
                            })}
                        </>
                        <Button
                            label={categoryIndex < categories.length - 1
                                ? content.step2ProceedButton2
                                : content.step2ProceedButton3}
                            type="submit"
                            onClick={nextStep} />
                    </>
                )}
        </div>
    );
}
