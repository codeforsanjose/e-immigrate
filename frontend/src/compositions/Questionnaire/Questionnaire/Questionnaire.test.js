import React from 'react';
import Questionnaire from './Questionnaire';

import renderer from 'react-test-renderer';

it('Questionnaire renders correctly', () => {
    const tree = renderer
        .create(
            <Questionnaire
                filteredQuestions={[]}
                bindField={() => {}}
                questions={[]}
                setAllFieldsTouched={() => {}}
                submitQuestionnaireResponse={() => {}}
                questionnaireResponse={() => {}}
                setQuestionnaireResponse={() => {}}
            />
        )
        .toJSON();
    expect(tree).toMatchSnapshot();
});
