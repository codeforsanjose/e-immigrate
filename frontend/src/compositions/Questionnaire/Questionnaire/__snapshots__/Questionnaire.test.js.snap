// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Questionnaire renders correctly 1`] = `
<div
  className="Questionnaire"
>
  <div
    className="QuestionnaireIntro"
  >
    <div>
      <h1>
        :
      </h1>
      <h2 />
      <ul>
        <li />
        <li />
      </ul>
    </div>
    <div>
      <h1 />
      <ul>
        <li />
        <li />
        <li />
      </ul>
    </div>
    <div
      className="ButtonContainer"
    >
      <button
        className="buttonComponent"
        onClick={[Function]}
      />
    </div>
  </div>
</div>
`;
